src/context/blocked.rs

```diff
@@ -1,18 +1,151 @@
+    let password = ValkeyString::new(Some(ctx_ptr), password);
+
+    // Get private data from context
+    let priv_data = context.get_blocked_client_privdata();
+    if priv_data.is_null() {
+        context.log_warning("[auth_reply_wrapper] Private data is null; this should not happen!");
+        return AUTH_NOT_HANDLED;
+    }
+
+    let callbacks = &*(priv_data
+        as *const (
+            AuthReplyCallback<T>,
+            Option<FreePrivDataCallback<T>>,
+            Box<T>,
+        ));
+
+    match (callbacks.0)(&context, username, password, callbacks.2.as_ref().clone()) {
+        Ok(status) => status,
+        Err(error) => {
+            let error_msg = ValkeyString::create(Some(ctx_ptr), error.to_string().as_str());
+            *err = error_msg.into_raw();
+            AUTH_HANDLED
+        }
+    }
+}
+
+#[allow(dead_code)]
+unsafe extern "C" fn free_callback_wrapper<T: 'static>(
+    ctx: *mut raw::RedisModuleCtx,
+    privdata: *mut c_void,
+) {
+    let context = Context::new(ctx);
+
+    if privdata.is_null() {
+        context
+            .log_warning("[free_callback_wrapper] Private data is null; this should not happen!");
+        return;
+    }
+
+    let (_, free_cb, data) = *Box::from_raw(
+        privdata
+            as *mut (
+                AuthReplyCallback<T>,
+                Option<FreePrivDataCallback<T>>,
+                Box<T>,
+            ),
+    );
+
+    // free_cb is optional, so we only call free_cb if it exists
+    // If no free_cb exists but the data(priv_data) exists, that's fine -
+    // user might not allocate memory using VM_ALLOC or VM_REALLOC hence
+    // it is not explicitly needed to free using VM_FREE and rust implicit
+    // drop will handle cleanup. If memory was allocated using VM_ALLOC,
+    // the module owners are responsible to provide free_privdata_callback
+    // to avoid memory leaks.
+    if let Some(free_cb) = free_cb {
+        free_cb(&context, *data);
+    }
 }
 
 // We need to be able to send the inner pointer to another thread
-unsafe impl Send for BlockedClient {}
+unsafe impl<T> Send for BlockedClient<T> {}
 
-impl Drop for BlockedClient {
+impl<T> BlockedClient<T> {
+    pub(crate) fn new(inner: *mut raw::RedisModuleBlockedClient) -> Self {
+        Self {
+            inner,
+            auth_reply_callback: None,
+            free_priv_callback: None,
+            private_data: None,
+        }
+    }
+
+    #[allow(dead_code)]
+    pub(crate) fn with_auth_callback(
+        inner: *mut raw::RedisModuleBlockedClient,
+        auth_reply_callback: AuthReplyCallback<T>,
+        free_priv_callback: Option<FreePrivDataCallback<T>>,
+    ) -> Self {
+        Self {
+            inner,
+            auth_reply_callback: Some(auth_reply_callback),
+            free_priv_callback,
+            private_data: None,
+        }
+    }
+
+    pub fn set_blocked_private_data(&mut self, data: T) {
+        self.private_data = Some(Box::new(data));
```

Comment:
Question: If someone were to set private data on a Blocked Client (General blocking, not Auth), will they leak memory? We do not currently accept a free callback for the General Blocking case, so the Module cannot free it.

This will be addressed once we extend this structure to accept free_priv callbacks for the General Blocking case as well.

Do you want to include general blocking in a separate PR or this PR?

If we are extending this structure to handle general blocking (that accepts reply and free callbacks) in a separate PR, then we should add a documentation comment (`///`) saying this should only be used for blockonauth until this interface supports general blocking